### PR TITLE
Parse Physical Chassis

### PR DESCRIPTION
**Pull request by caiocmpaes**
- Parse Physical Chassis informations and maps the relationship to Physical Servers that live inside a chassi.
- Added chassis to the /cabinet mock.

Depends on:
- ~[ManageIQ/manageiq-schema#183](https://github.com/ManageIQ/manageiq-schema/pull/183) [Merged]~

### CODE REVIEW

**Comment by rodneyhbrown7** [comment]
The plural and singular spellings of chassis are the same. Can you change all instances of 'chassi'  or 'Chassi' to 'chassis' or 'chassis'?

**Comment by caiocmpaes** [comment]
@miq-bot remove_label unmergeable

File: app/models/manageiq/providers/lenovo/physical_infra_manager/parsers/components/physical_chassis_parser.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
+require_relative 'component_parser'
+
+module ManageIQ::Providers::Lenovo
+  module Parsers
+    class PhysicalChassisParser < ComponentParser
+      class << self
+        # parse a node object to a hash with physical chassis data
+        # @param [XClarityClient::Node] node - object containing physical chassis data
```
Comment (douglasgabriel) [inline app/models/manageiq/providers/lenovo/physical_infra_manager/parsers/components/physical_chassis_parser.rb]: I think that you will want to change the `@param` to its new name `chassi_hash` and add some comment about the `rack` param.

File: app/models/manageiq/providers/lenovo/physical_infra_manager/parsers/components/physical_server_parser.rb [hunk 2]
```diff
@@ -13,12 +13,13 @@ class << self
         #
         # @return [Hash] containing the physical server information
         #
-        def parse_physical_server(node, rack = nil)
+        def parse_physical_server(node_hash, rack = nil, chassis = nil)
```
Comment (douglasgabriel) [inline app/models/manageiq/providers/lenovo/physical_infra_manager/parsers/components/physical_server_parser.rb]: Please, also update the param name on doc comment

Comment (douglasgabriel) [reply app/models/manageiq/providers/lenovo/physical_infra_manager/parsers/components/physical_server_parser.rb]: Would be good add something about the new params `rack` and `chassis`, too

File: app/models/manageiq/providers/lenovo/physical_infra_manager/physical_chassis.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+module ManageIQ::Providers
+  class Lenovo::PhysicalInfraManager::PhysicalChassis < ::PhysicalChassis
+    delegate :product_name,
```
Comment (douglasgabriel) [inline app/models/manageiq/providers/lenovo/physical_infra_manager/physical_chassis.rb]: I'm wondering if this `delegate` strategy is good here, seems like it is a good thing when you want to support some code that already access this properties like `chassis.product_name`, for example, which is not the case once it is a new component. What do you think @agrare , is there some pattern about this strategy?

File: spec/models/manageiq/providers/lenovo/physical_infra_manager/refresh_parser_spec.rb [hunk 4]
```diff
@@ -36,16 +36,68 @@
...
+      VCR.use_cassette("#{described_class.name.underscore}_ems_inv_to_hashes") do
+        ems_inv_to_hashes
+      end
+    end
+
+    it 'will retrieve physical all chassis' do
+      expect(result[:physical_chassis].size).to eq(1)
+    end
+
+    it 'will parse physical chassi fields' do
```
Comment (douglasgabriel) [inline spec/models/manageiq/providers/lenovo/physical_infra_manager/refresh_parser_spec.rb]: I think that you wanted to write `chassis` here

File: spec/models/manageiq/providers/lenovo/physical_infra_manager/refresh_parser_spec.rb [hunk 5]
```diff
@@ -36,16 +36,68 @@
...
+      physical_chassi = result[:physical_chassis].first
```
Comment (douglasgabriel) [inline spec/models/manageiq/providers/lenovo/physical_infra_manager/refresh_parser_spec.rb]: Also here should be `physical_chassis`

File: spec/models/manageiq/providers/lenovo/physical_infra_manager/refresh_parser_spec.rb [hunk 6]
```diff
@@ -36,16 +36,68 @@
...
+      expect(asset_detail[:room]).to eq("")
+      expect(asset_detail[:rack_name]).to eq("")
+      expect(asset_detail[:lowest_rack_unit]).to eq(0)
+    end
+
+    it 'will have a reference to the physical rack where the chassis is inside' do
+      physical_rack = result[:physical_racks].first
+      physical_chassis = result[:physical_chassis].first
+
+      expect(physical_chassis[:physical_rack][:ems_ref]).to eq(physical_rack[:ems_ref])
```
Comment (douglasgabriel) [inline spec/models/manageiq/providers/lenovo/physical_infra_manager/refresh_parser_spec.rb]: Seems like this is a redundant test if you are verifying the equality between `physical_chassis[:physical_rack]` and `physical_rack[:ems_ref]` as you do bellow.

File: spec/models/manageiq/providers/lenovo/physical_infra_manager/refresh_parser_spec.rb [hunk 7]
```diff
@@ -90,32 +142,47 @@
...
       hardware = computer_system[:hardware]
 
       expect(hardware[:disk_capacity]).to be_nil
     end
+
+    it 'will have a reference to the physical chassis where the server is inside' do
+      physical_server = @result[:physical_servers][1]
+      physical_chassis = @result[:physical_chassis][0]
+
+      expect(physical_server[:physical_chassis][:ems_ref]).to eq(physical_chassis[:ems_ref])
```
Comment (douglasgabriel) [inline spec/models/manageiq/providers/lenovo/physical_infra_manager/refresh_parser_spec.rb]: Same redundancy that I commented before :smile:

**Comment by caiocmpaes** [comment]
@miq-bot remove_label unmergeable

**Comment by caiocmpaes** [comment]
@miq-bot remove_label wip

**Comment by caiocmpaes** [comment]
@douglasgabriel I fixed all the corrections that you pointed. Thank you!

Also, I rebased my code after the refactoring of "relative requires".

**Comment by agrare** [comment]
@caiocmpaes can you rebase this PR?

**Comment by douglasgabriel** [comment]
Thanks @caiocmpaes , LGTM now :+1: Only needing the rebase that @agrare mentioned

**Comment by caiocmpaes** [comment]
Rebased @agrare

**Comment by caiocmpaes** [comment]
@miq-bot remove_label unmergeable

**Comment by agrare** [comment]
@caiocmpaes this changeset still does not apply cleanly

**Comment by caiocmpaes** [comment]
Now Its is rebased again, @agrare

**Pull request merged**